File: clojure/bob/bob.clj
;;;; Bob is a lackadaisical teenager. In conversation, his responses
;;;; are very limited.
;;;;
;;;; Bob answers 'Sure.' if you ask him a question.
;;;;
;;;; He answers 'Woah, chill out!' if you yell at him.
;;;;
;;;; He says 'Fine. Be that way!' if you address him without actually
;;;; saying anything.
;;;;
;;;; He answers 'Whatever.' to anything else.

(ns bob)

(defn- question? [str]
  "Return true if str ends with a ?"
  (= (last str) \?))

(defn- yelling? [str]
  "Return true if the string only contains upper-case;
  except when string contains only numbers, in which case,
  return false"
  (and 
    (= str (clojure.string/upper-case str))
    (re-seq #"[a-zA-Z]" str)))

(defn response-for [str]
  (cond
    (clojure.string/blank? str) "Fine. Be that way!"
    (yelling? str) "Woah, chill out!"
    (question? str) "Sure."
    :else "Whatever."))

File: haskell/bob/Bob.hs
module Bob ( responseFor ) where
import Data.Char

isQuestion :: String -> Bool
isQuestion que = last que == '?'

isShouting :: String -> Bool
isShouting str = any isAlpha str && not (any isLower str)

isSilence :: String -> Bool
isSilence = all isSpace

responseFor :: String -> String
responseFor x
    | isSilence x = "Fine. Be that way!"
    | isShouting x = "Woah, chill out!"
    | isQuestion x = "Sure."
    | otherwise = "Whatever."
